types/json/binary.go

```diff
@@ -112,6 +112,14 @@ func (bj BinaryJSON) String() string {
 	return string(out)
 }
 
+// ToHashKey generate hash key using bj.TypeCode and bj.Value
+func (bj *BinaryJSON) ToHashKey() ([]byte, error) {
```

Comment:
This implementation might be incorrect, since JSON does not do byte comparing. For example, the compare implementation discard the epsilon when comparing JSON floats while using this hash function they will be different. Also due to this reason [it might even not possible to write a correct hash function](https://softwareengineering.stackexchange.com/questions/391100/how-to-implement-float-hashing-with-approximate-equality) for JSON.

Reply:
Thank you for pointing out, neither for types.ETReal?

Reply:
No, only equality of json depends on an epsilon. Only json cannot be optimized by hash lookup.

(An ordered map lookup_set is ok

Reply:
It seems need the equality of float also need depend on an epsilon. See https://github.com/pingcap/tidb/pull/13619

Reply:
I can delete the implementation for json IN for this issue https://github.com/pingcap/tidb/issues/13710#issuecomment-564983426
And I can make a sorted array for ETReal and use binary search.

Reply:
@SunRunAway MySQL `WON'T` compare with epsilon for real numbers:

```
mysql> create table t(a double, b double, c double);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (0.1, 0.2, 0.3);
Query OK, 1 row affected (0.01 sec)

mysql> select * from t where a+b=c;
Empty set (0.00 sec)
```

Reply:
try `create table tf(a double(9,4), b double(9,4), c double(9,4));`

Reply:
```
mysql> create table t(a double(9,4), b double(9,4), c double(9,4));
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (0.1, 0.2, 0.0);
Query OK, 1 row affected (0.02 sec)

mysql> update t set c=a+b;
Query OK, 1 row affected (0.01 sec)
Rows matched: 1  Changed: 1  Warnings: 0

mysql> select * from t;
+--------+--------+--------+
| a      | b      | c      |
+--------+--------+--------+
| 0.1000 | 0.2000 | 0.3000 |
+--------+--------+--------+
1 row in set (0.00 sec)

mysql> select * from t where c = 0.3;
+--------+--------+--------+
| a      | b      | c      |
+--------+--------+--------+
| 0.1000 | 0.2000 | 0.3000 |
+--------+--------+--------+
1 row in set (0.00 sec)

mysql> drop table t;
Query OK, 0 rows affected (0.01 sec)

mysql> create table t(a double, b double, c double);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (0.1, 0.2, 0.0);
Query OK, 1 row affected (0.01 sec)

mysql> update t set c=a+b;
Query OK, 1 row affected (0.00 sec)
Rows matched: 1  Changed: 1  Warnings: 0

mysql> select * from t;
+------+------+---------------------+
| a    | b    | c                   |
+------+------+---------------------+
|  0.1 |  0.2 | 0.30000000000000004 |
+------+------+---------------------+
1 row in set (0.00 sec)

mysql> select * from t where c = 0.3;
Empty set (0.00 sec)
```

Again it cannot prove that MySQL compares with eps, since the result of a+b in (9, 4) precision is **precisely** equal to c.

Reply:
Mysql uses precesion to dectemine an epsilon, see https://github.com/mysql/mysql-server/blob/8.0/sql/item_cmpfunc.cc#L794-L796, same as in PR #13619 (https://github.com/pingcap/tidb/pull/13619/files#diff-cac1281ffeb176bb1b250418779849d6R2455-R2466)